src/frequenz/sdk/timeseries/_ringbuffer/buffer.py

```diff
@@ -154,10 +165,14 @@ def update(self, sample: Sample[QuantityT]) -> None:
         )
 
         # Update data
-        value: float = np.nan if sample.value is None else sample.value.base_value
+        if self.has_value(sample):
+            assert sample.value is not None
```

Comment:
I just noticed this, too bad that this is needed, because it was just checked by `has_value()`.

You could avoid this by making `has_value` a function instead and using `TypeGuard`, but that would also need to take it back to operate on `float | None`:

```py
from typing import TypeGuard

def has_value(sample: float | None) -> TypeGuard[float]:
    """Check if a sample has a value and it's not NaN.
    Args:
        sample: sample to check.
    Returns:
        True if the sample has a value and it's not NaN.
    """
    return value is not None and not value.isnan()

if has_value(sample):  # If this returns True, then the type checker narrows the type to `float`
    value = sample.value.base_value  # this is now fine without the assert
else:
    value = np.nan
```

I thought about alternatives and all seem awful, so I guess we'll have to live with the assert :shrug: